src/Orleans.Runtime/Networking/SiloConnectionFactory.cs

```diff
@@ -32,6 +35,15 @@ internal sealed class SiloConnectionFactory : ConnectionFactory
             this.localSiloDetails = localSiloDetails;
         }
 
+        public override ValueTask<Connection> ConnectAsync(SiloAddress address, CancellationToken cancellationToken)
+        {
+            EnsureInitialized();
+
+            if (this.siloStatusOracle.IsDeadSilo(address)) throw new ConnectionAbortedException($"Denying connection to known-dead silo {address}");
```

Comment:
This check is performed here just in case there are multiple messages already waiting for a connection. This lets us largely avoid a race, eg when a silo is declared dead while several tasks are already waiting to get a `Connection` to send it a message.

There is still a case where the silo might be declared dead while a particular connection attempt is pending. We could theoretically fix but I don't think the added complexity would be worthwhile: the connection attempt will timeout fairly quickly and fail and retries will fail-fast.